src/websockets/index.js

```diff
@@ -118,64 +73,66 @@ export default async (expServer) => {
+
   wss.on("connection", async (ws, req) => {
     ws.isAlive = true;
+    const { userId } = req;
+
     ws.on("pong", () => {
       ws.isAlive = true;
     });
 
-    const { userId } = req;
-
     insertNewUser(connections, userId, ws);
 
-    /*
+    /**
      * Here we get the timer from memory or from the database and send it to the client.
-     * We don't broadcast it
      */
-    await getTimer(clients, userId);
-    ws.send(await handleMessage(action.GET_TIMER, clients, userId));
+    const clientTimer = await getClient(clients, userId);
+    ws.send(JSON.stringify(clientTimer));
 
-    /*
-    Here we handle the messages from the client.
-    And we broadcast the response to all the clients that are connected to the same user.
+    /**
+    * Here we handle the messages from the client.
+    * And we broadcast the response to all the clients that are connected to the same user.
     */
     ws.on("message", async (data) => {
       const resp = await handleMessage(data, clients, userId);
       broadcastToSameUser(connections, userId, resp);
     });
 
-    /*
-    Here we handle the close event.
-    If there is another connection to the same user, we don't do anything.
-    Else he is the last connection and we do a forced pause if need be.
-    This may happen if the user closes all the tabs or the browser or he lost connection with
-    the service
-    We then remove the connection from the connections map.
+    /**
+    * Here we handle the close event.
+    * If there is another connection to the same user, we don't do anything.
+    * Else he is the last connection and we do a forced pause if need be.
+    * This may happen if the user closes all the tabs or the browser or he lost connection with
+    * the service
+    * We then remove the connection from the connections map.
     */
     ws.on("close", async () => {
-      if (!checkOtherConn(connections, ws, userId)) {
-        await handleMessage(action.FORCED_PAUSE, clients, userId);
+      if (!hasOtherConn(connections, userId, ws)) {
+        const client = clients.get(userId);
+        if (client.started && !client.paused) {
+          await handleMessage(action.FORCED_PAUSE, clients, userId);
+        }
       }
       removeConnection(connections, userId, ws);
     });
   });
 
-  // The function to check if the connection is alive
-  const interval = setInterval(async () => {
-    wss.clients.forEach(async (ws) => {
-      if (ws.isAlive === false) return ws.terminate();
-
+  // For each new connection we start a time interval of 1min to check if the connection is alive.
+  // change to 1min before push
+  const interval = setInterval(() => {
+    wss.clients.forEach((ws) => {
+      if (ws.isAlive === false) {
+        return ws.close();
+      }
       ws.isAlive = false;
       ws.ping();
     });
-  }, 3000000);
+  }, 10000);
 
-  // Here we just clear the interval when the server closes
-  wss.on("close", () => {
+  wss.on('close', () => {
```

Comment:
Nice work.

However, can it be modified to automatically save the last state of the user's timer when the connection has been terminated; such that when the user logs in again; he/she resumes from their last state.

Reply:
Thank you for your advice! The feature you mentioned is implemented in the clientHandler file, in which every changes would be saved to the database.

Reply:
Awesome.